juju/juju
# feat: model migrate authorized keys

This PR introduces the ability model migrate authorized keys for a model between controllers. We are doing this as authorized keys are no longer apart of model config in Juju 4.0.  To achieve this we have updated the description package with a new first class type `AuthorizedKeys`. This allows 4.0 and greater controllers to transfer authorized keys via the new attribute. For the case of migrating 3.6 to 4.0 we are still pulling the keys off of model config.  This has been done because introduce transformation logic to the description package moves too much business logic out of Juju's control.

As part of this PR I have also added bash tests to check migration works correctly. Various other little cleanup tasks have been performed as part of this PR to move some constants around.

## Checklist


## QA steps

To test the 4.0 case you can `cd tests; ./main.sh -v authorized_keys test_migrate_authorized_keys`

For the 3.6 to 4.0 case this still has to be done manually:

```bash
juju36 bootstrap localhost from
juju36 add-model keys
juju36 add-ssh-key
juju4 bootstrap localhost to
juju migrate from:keys to
juju switch keys
juju list-ssh-keys # Confirm your key exists
```

## Documentation changes

N/A

## Links

**Jira card:** JUJU-6280
